docs/community/rust-coding-guidelines/rust-secure-coding-guidelines.md

```diff
@@ -0,0 +1,253 @@
+Both traits are _unsafe traits_, i.e., the Rust compiler does not verify in any way that they are implemented correctly. The danger is real: an incorrect implementation may lead to **undefined behavior**.
+
+In the majority of scenarios, manual implementation is unnecessary. In Rust, nearly all primitive types intrinsically implement Send and Sync traits, and for a significant proportion of compound types, the Rust compiler automatically derives these implementations.
+
+### Comparison Traits
+
+Ensure the implementation of standard comparison traits respects documented invariants [[comparison traits partialeq eq partialord ord]](https://anssi-fr.github.io/rust-guide/06_typesystem.html#comparison-traits-partialeq-eq-partialord-ord).
+
+> The ANSSI resource extensively covers the matter, an example in [Appendix](#implementation-of-comparison-traits) is provided as a really basic demonstration of implementation.
+
+### Enums
+
+Prefer enums for state management to prevent invalid state representation.
+
+### Concurrency Safe Primitives
+
+Make use of Rust’s concurrency primitives like `Arc`, `Mutex`, and `RwLock` to manage shared state [[Rustbook: concurrency]](https://doc.rust-lang.org/book/ch16-00-concurrency.html).
+By utilizing these primitives, Rust programs can manage shared resources among multiple threads safely and efficiently, adhering to Rust's goals of enabling fearless concurrency and memory safety. This is critical in multithread programs because mistakes in concurrency can be very costly, both for the system's stability and security.
+
+### Data Structures with Deterministic Internal Order
+
+Certain data structures, like HashMap and HashSet, do not guarantee a deterministic order for the elements stored within them. This lack of order can lead to problems in operations that require processing elements in a consistent sequence across multiple executions. Below is a list of deterministic data structures available in Rust. Please note, this list may not be exhaustive:
+
+- **BTreeMap:** It maintains its elements in sorted order by their keys.
+- **BinaryHeap:** It maintains its elements in a heap order, which is a complete binary tree where each parent node is less than or equal to its child nodes.
+- **VEC**: It maintains its elements in the order in which they were inserted. ⚠️
+- **LinkedList:** It maintains its elements in the order in which they were inserted. ⚠️
+- **VecDeque:** It maintains its elements in the order in which they were inserted. ⚠️
+
+## Cryptography
+
+### Talk to a cryptographer
+
+If you’re using cryptography in your code and you are **not** a cryptographer, you don’t know what you don’t know (_e.g.,_ small subgroup attacks, malleability attacks, domain-separation attacks, collision resistance versus one-wayness, safe key reuse, correctly using randomness, nonce reuse, CCA security, CPA security, etc). So ask for help from a cryptographer in our team (Alin, Benny, Michael, Rex).
```

Comment:
this guidance is not actionably for anyone that doesn't work at aptos labs. these guidelines are public and should be applicable to everonye. I would remove this.

Reply:
I think this came from someone else, I noticed the change while porting, but left while thinking for a better rephrasing.